File: helium-java/src/main/java/com/heliumhq/util/XPath.java
package com.heliumhq.util;

import static com.heliumhq.util.StringUtils.isEmpty;
import static java.lang.String.format;

public class XPath {
	public static String lower(String text) {
		String alphabet =
			"ABCDEFGHIJKLMNOPQRSTUVWXYZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝ";
		return format(
			"translate(%s, '%s', '%s')", text, alphabet, alphabet.toLowerCase()
		);
	}
	public static String replaceNbsp(String text) {
		return replaceNbsp(text, " ");
	}
	public static String replaceNbsp(String text, String by) {
		return format("translate(%s, '\u00a0', '%s')", text, by);
	}
	public static String predicate(String condition) {
		if (isEmpty(condition))
			return "";
		else
			return format("[%s]", condition);
	}
	public static String predicateOr(String... conditions) {
		String resultCondition = "";
		for (String condition : conditions) {
			if (!isEmpty(condition)) {
				if (!isEmpty(resultCondition))
					resultCondition += " or ";
				resultCondition += condition;
			}
		}
		return predicate(resultCondition);
	}
}

File: helium/_impl/util/xpath.py
# -*- coding: utf-8 -*-
def lower(text):
	alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝ'
	return "translate(%s, '%s', '%s')" % (text, alphabet, alphabet.lower())

def replace_nbsp(text, by=' '):
	return "translate(%s, '\u00a0', %r)" % (text, by)

def predicate(condition):
	return '[%s]' % condition if condition else ''

def predicate_or(*conditions):
	return predicate(' or '.join([c for c in conditions if c]))